sql/service_wsrep.cc

```diff
@@ -302,3 +320,13 @@ extern "C" void wsrep_commit_ordered(THD *thd)
     thd->wsrep_cs().ordered_commit();
   }
 }
+
+extern "C" bool wsrep_thd_set_wsrep_killed(THD *thd)
+{
+  if (thd->wsrep_killed)
+  {
+    return true;
+  }
+  thd->wsrep_killed = true;
+  return false;
+}
```

Comment:
In `wsrep_thd_bf_abort()`, `thd->wsrep_killed` is protected by `thd->LOCK_thd_data`. Can we protect it also here? In the current state, we can have a race condition between multiple `wsrep_thd_set_wsrep_killed()` executions.

If using a mutex is not possible, could we at least use `Atomic_wrapper` for `wsrep_killed`, and perform the test-and-set as an atomic operation?

The assignment operator in MariaDB code should not be preceded by a space:
```c++
thd->wsrep_killed= true;
```

Reply:
I have not yet thoroughly reviewed this but if we at this point own lock_sys->mutex and trx->mutex(victim) LOCK_thd_data is exactly something we should not take, it just moves deadlock on different location.

Reply:
Only place where this is called already has locked LOCK_thd_data